### PR TITLE
File upload Action 

### PR DESCRIPTION
**Pull request by snehmehta**
Ticket : https://github.com/EnsembleUI/ensemble/issues/129

Sample YAML configuration
```yaml
Button:
  label: Pick
  onTap:
    action: uploadFiles
    id: filePicker
    url: https://5004-103-238-108-177.ngrok.io/core/file/multiple_upload/
    options: 
      allowMultiple: True
      allowExtensions: 
        - jpg
        - pdf
    onComplete: |
      //@code
      ensemble.debug(filePicker.files[0]['name'])
      
    onError: |
      //@code
      ensemble.debug('Failed to upload files')

```

### CODE REVIEW

File: lib/util/utils.dart [hunk 1]
```diff
@@ -298,7 +298,17 @@ class Utils {
           payload['url'],
           openInExternalApp: Utils.getBool(payload['openInExternalApp'], fallback: false)
         );
-      }
+      } else if ( payload['action']  == ActionType.uploadFiles.name ) {
+         return FileUploadAction(
+            id: Utils.optionalString(payload['options']['id']),
```
Comment (vusters) [inline lib/util/utils.dart]: 'options' is optional. This will throw an error. Same as below

File: lib/screen_controller.dart [hunk 2]
```diff
@@ -310,6 +310,23 @@ class ScreenController {
...
       launchUrl(Uri.parse(value),mode: (action.openInExternalApp)?LaunchMode.externalApplication:LaunchMode.platformDefault );
+    } else if (action is FileUploadAction) {
+      FilePicker.platform.pickFiles(
+        type: action.allowedExtensions == null ? FileType.any: FileType.custom,
+        allowedExtensions: action.allowedExtensions,
+        allowCompression: action.allowCompression ?? true,
+        allowMultiple: action.allowMultiple ?? false,
+      ).then((files) {
+        
+        if (!(files?.files.isNotEmpty ?? false)) return;
```
Comment (vusters) [inline lib/screen_controller.dart]: cool but not readable. Just use files==null || files?.files.isEmpty please.

File: lib/screen_controller.dart [hunk 3]
```diff
@@ -310,6 +310,23 @@ class ScreenController {
...
+        if (action.url == null) throw Exception('Enter URL');
```
Comment (vusters) [inline lib/screen_controller.dart:327]: use LanguageException so it shows up in the Preview UI.

File: lib/util/upload_utils.dart [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      }
+    } catch (error) {
+      onError?.call(error);
+    }
+    
+  }
+}
+
+
+class MultipartRequest extends http.MultipartRequest {
```
Comment (vusters) [inline lib/util/upload_utils.dart:54]: This is cool. What's the requirement on the server side to accept files?

Comment (snehmehta) [reply lib/util/upload_utils.dart:54]: Requirements.
- API method is expected to be POST and accepts content type multipart/form-data

Side note
To access file server code can use either `request.files` or `request.data['files']`

**Comment by vusters** [comment]
This looks good. Can you add this definition to ensemble-schema.json so typeahead works in the Studio. See any existing widgets in there for reference.

**Comment by kmahmood74** [comment]
@snehmehta the ensemble schema is here - https://github.com/EnsembleUI/ensemble/blob/main/assets/schema/ensemble_schema.json

Kindly wrap this up

File: assets/schema/ensemble_schema.json [hunk 5]
```diff
@@ -1482,6 +1482,51 @@
...
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "id": {
+                "type": "string",
+                "description": "Give the Uploader an ID allows you to bind to its result. e.g. ${filePicker.files...}, ${filePicker.body...}"
+              },
+              "url": {
```
Comment (vusters) [inline assets/schema/ensemble_schema.json:1500]: can you change this to uploadUrl here and the code to be more explicit?

**Review by vusters — APPROVED** [review]

**Comment by kmahmood74** [comment]
@snehmehta there is a conflict, after resolving that, feel free to merge and close

**Pull request merged**